### PR TITLE
chore: Replace fast sync with batch processing

### LINKED ISSUE #118: Cherry-pick batch submission to replace fast sync
* Cherry-pick the batch submission change https://github.com/babylonlabs-io/finality-provider/pull/57 or do it from scratch
* Remove fast sync completely

### PR DESCRIPTION
**Pull request by gitferry**
Closes #118. Notable changes:
* remove fast sync completely
* introduce batch process for catching up
* introduce signature submission interval to config
* introduce batch submission size to config

### CODE REVIEW

File: finality-provider/service/fp_instance.go
Comment (Lazar955) [inline finality-provider/service/fp_instance.go:713]: nit: wrap errors

**Review by Lazar955 — APPROVED** [review]
lgtm! Good job simplifying the logic!

File: finality-provider/config/config.go
Comment (bap2pecs) [inline finality-provider/config/config.go:38]: Shouldn't we add a config for batch processing size? what's the current code behavior if it lags way behind?

File: finality-provider/config/poller.go
Comment (bap2pecs) [inline finality-provider/config/poller.go:7]: i think the polling interval shouldn't be less than the chain block-producing time. OP chain is usually 2-3 seconds. what's the block time for Babylon chain?

File: finality-provider/service/fp_instance.go
Comment (bap2pecs) [inline finality-provider/service/fp_instance.go:209]: there might be a problem here since `shouldProcessBlock` makes RPC calls to the Babylon chain. so in this `for` loop we are doing a lot of RPC calls.

File: finality-provider/service/fp_instance.go
Comment (bap2pecs) [inline finality-provider/service/fp_instance.go:337]: we can probably get rid of `SubmitFinalitySignature` from the interface

File: finality-provider/service/fp_instance.go
Comment (bap2pecs) [inline finality-provider/service/fp_instance.go:794]: nit: rename to `getFinalityActivationHeightWithRetry`

File: finality-provider/service/fp_instance.go
Comment (gitferry) [reply finality-provider/service/fp_instance.go:337]: I would keep the interface but the implementation could just use the one from the batch submission

File: finality-provider/service/fp_instance.go
Comment (gitferry) [reply finality-provider/service/fp_instance.go:209]: Agreed this case could happen in case of catching up. Left a TODO here for now

File: finality-provider/config/poller.go
Comment (gitferry) [reply finality-provider/config/poller.go:7]: I would argue that we should have a short polling interval for a quick catching up. Would `ms` level interval a concern in practice?

File: finality-provider/config/config.go
Comment (gitferry) [reply finality-provider/config/config.go:39]: Good point. Fixed.

Comment (gitferry) [reply finality-provider/config/config.go:39]: If the fp lags behind, the poller should quickly poll blocks (that's why we should have a short polling interval). Then the fp will do batch submission

**Comment by gitferry** [comment]
I would wait for one more approval before merge

File: finality-provider/config/poller.go
Comment (bap2pecs) [reply finality-provider/config/poller.go:7]: Catching up is not a normal case. If it's way behind, people should just change the config file temporarily to speed up. imo default value should be for normal  submission cases

**Review by bap2pecs — APPROVED** [review]
lgtm. can we cherrypick this to the base consumer branch?

File: finality-provider/config/poller.go
Comment (gitferry) [reply finality-provider/config/poller.go]: Makes sense. Changed it to `1s` and added notes there

**Comment by gitferry** [comment]
> lgtm. can we cherrypick this to the base consumer branch?

Sure.

File: finality-provider/service/fp_instance.go
Comment (KonradStaniec) [inline finality-provider/service/fp_instance.go:207]: I think you could also add:
```
case <-fp.quit:
```

to this select, to not wait long time for quit in case of many blocks pending on poller channel.

**Review by KonradStaniec — APPROVED** [review]
looks much cleaner now  👍

**Comment by bap2pecs** [comment]
@gitferry do you plan to backport this to the base branch?

**Comment by gitferry** [comment]
@bap2pecs not sure it's efficient to backport every change to the base branch. Do we have a routine to periodically backport changes to the base branch?

**Comment by bap2pecs** [comment]
> @bap2pecs not sure it's efficient to backport every change to the base branch. Do we have a routine to periodically backport changes to the base branch?

actually it's probably the opposite. I feel it's probably always better to backport right away or context will be lost. for example, https://github.com/babylonlabs-io/finality-provider/pull/153 is not a straightforward backport due to some interface level difference between main and base branch. but I can quickly backport it b/c my memory was fresh so I know where to change

also, i'd say for important changes like this, we should proritize. this one is currently blocking some of our works (e.g. e2e test refactoring) so we'd appreciate if it can be backported

**Comment by bap2pecs** [comment]
> @bap2pecs Make sense. I'll backport it

thanks a lot! I remember @vitsalis   told me in Bangkok that this backport issue will be solved soon but i don't remember the conversation

do you guys feel that backporting takes lots of engineering time for your team? @gitferry @SebastianElvis

**Pull request merged**
